sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala

```diff
@@ -129,3 +138,17 @@ case class UserDefinedFunction protected[sql] (
     }
   }
 }
+
+// We have to use a name different than `UserDefinedFunction` here, to avoid breaking the binary
+// compatibility of the auto-generate UserDefinedFunction object.
+private[sql] object SparkUserDefinedFunction {
```

Comment:
Oh, so we're working around by new object here. SGTM. I assumed mutable status and copy stuff but trivial or not an issue.